mathics/builtin/numbers/calculus.py

```diff
@@ -2254,8 +2255,7 @@ def eval(self, eqs, vars, evaluation: Evaluation):
             elif eq is SymbolFalse:
                 return ListExpression()
             elif not eq.has_form("Equal", 2):
-                evaluation.message("Solve", "eqf", eqs)
-                return
+                sympy_conditions.append(eq.to_sympy())
```

Comment:
It must be checked that eq is a condition, and has a sympy form.

Reply:
The test `Solve[{x^2==4 && x < 0 && x > -4},{x}]` fails because here the first argument is treated as a single condition `x^2==4 && x < 0 && x > -4` without any equation. Conjunctions should be decomposed to work.

Reply:
> It must be checked that eq is a condition, 

It's my bad, I'll try to investigate the way to do this.

> and has a sympy form.

I noticed this issue while writing these lines, but I think this problem is out of the scope of this contribution. I have little knowledge over Mathics, so if I said anything wrong please do point out. My argument is that anything that to_sympy() can't translate to a sympy expression shouldn't be considered as valid. In this case, the exception would catch the error and fail the expression.  

> The test `Solve[{x^2==4 && x < 0 && x > -4},{x}]` fails because here the first argument is treated as a single condition `x^2==4 && x < 0 && x > -4` without any equation. Conjunctions should be decomposed to work.

Agreed. Actually, I noticed this issue too while writing & testing, and put this test case intentionally looking for your opinion. If you could try `Solve[a^2 - 3a + 2 == 0 && a > 1, a]` with my code, you will find it works perfectly. I think this is a parsing issue, the parser can't always decompose conjunctions correctly. In fact, I also noticed that the parser can't handle things like `<some eqs> && <condition1> || <condition2>` neither, which produces a non-sympy object. In any casaes, I think this problem is also out of the scope of this contribution.

Reply:
> @BlankShrimp the parser can't always decompose conjunctions correctly.

Aha, no doubt I have little knowledge over Mathics. There's nothing wrong for Mathics, it's my job to decompose the conjunction inside solve.eval(). Wait for my code.

Reply:
> Agreed. Actually, I noticed this issue too while writing & testing, and put this test case intentionally looking for your opinion. If you could try `Solve[a^2 - 3a + 2 == 0 && a > 1, a]` with my code, you will find it works perfectly. I think this is a parsing issue, the parser can't always decompose conjunctions correctly. In fact, I also noticed that the parser can't handle things like `<some eqs> && <condition1> || <condition2>` neither, which produces a non-sympy object. In any casaes, I think this problem is also out of the scope of this contribution.

@BlankShrimp : Indeed, your example works, but `Solve[{x^2==4 && x<0}, x]` (the conjuction inside a list) doesn't. To do it right, the interpretation of the first argument should be done recursively. 

This could be a way to do that: replace the loop over   `eq_list`  by
```
        def process_equation(eq):
            if eq is SymbolTrue:
                pass
            elif eq.has_form("And", None):
                for sub_eq in eq.elements:
                    process_equation(sub_eq)
            elif eq.has_form("Equal", 2):
                left, right = eq.elements
                left = left.to_sympy()
                right = right.to_sympy()
                if left is None or right is None:
                    return
                eq = left - right
                eq = sympy.together(eq)
                eq = sympy.cancel(eq)
                sympy_eqs.append(eq)
                numer, denom = eq.as_numer_denom()
                sympy_denoms.append(denom)
            else:
                cond_sympy = eq.to_sympy()
                if cond_sympy is None:
                    evaluation.message("Solve", "eqf", eqs)
                    raise ValueError(str(eq))
                sympy_conditions.append(cond_sympy)
            return None

        for eq in eq_list:
            if eq is SymbolFalse:
                return ListExpression()
            try:
                process_equation(eq)
            except ValueError:
                return None
```
With this change, and one on the test on the existence of equations (`if not sympy_eqs`  ->  `if not sympy_eqs and sympy_conditions`) all the tests pass.

Reply:
On first look, it feels like the purpose of the above is to collect sympy_conditions. And this feels like something that might come up other places as well, right?  Perhaps this should be put inside `mathics.core.convert.sympy` then 

Small thing instead of `if ... pass` , `if ... return None` is more direct.